extendr-api/src/scalar/rint.rs

```diff
@@ -0,0 +1,284 @@
+                    let f = $expr;
+                    if let Some(res) = f(lhs) {
+                        // Note that if res = i32::MIN, this will also be NA.
+                        return Rint::from(res);
+                    }
+                }
+                Rint::na()
+            }
+        }
+
+        impl $opname1 for &Rint {
+            type Output = Rint;
+
+            #[doc = $docstring]
+            fn $opname2(self) -> Self::Output {
+                if let Some(lhs) = (*self).into() {
+                    let f = $expr;
+                    if let Some(res) = f(lhs) {
+                        // Note that if res = i32::MIN, this will also be NA.
+                        return Rint::from(res);
+                    }
+                }
+                Rint::na()
+            }
+        }
+    };
+}
+
+// Generate binary ops for +, -, * and /
+gen_binop!(
+    Add,
+    add,
+    |lhs: i32, rhs| lhs.checked_add(rhs),
+    "Add two Rint values or an option of i32, overflows to NA."
+);
+gen_binop!(
+    Sub,
+    sub,
+    |lhs: i32, rhs| lhs.checked_sub(rhs),
+    "Subtract two Rint values or an option of i32, overflows to NA."
+);
+gen_binop!(
+    Mul,
+    mul,
+    |lhs: i32, rhs| lhs.checked_mul(rhs),
+    "Multiply two Rint values or an option of i32, overflows to NA."
+);
+gen_binop!(
+    Div,
+    div,
+    |lhs: i32, rhs| lhs.checked_div(rhs),
+    "Divide two Rint values or an option of i32, overflows to NA."
+);
+
+// Generate unary ops for -, !
+gen_unnop!(
+    Neg,
+    neg,
+    |lhs: i32| Some(-lhs),
+    "Negate a Rint value, overflows to NA."
+);
+gen_unnop!(
+    Not,
+    not,
+    |lhs: i32| Some(!lhs),
+    "Logical not a Rint value, overflows to NA."
+);
+
+impl std::iter::Sum for Rint {
+    /// Sum an integer iterator over Rint.
+    /// Yields NA on overflow of NAs present.
+    fn sum<I: Iterator<Item = Rint>>(iter: I) -> Rint {
+        iter.fold(Rint::from(0), |a, b| a + b)
+    }
+}
+
+impl PartialEq<i32> for Rint {
+    /// Compare a Rint with an integer. NA always fails.
+    fn eq(&self, other: &i32) -> bool {
+        !self.is_na() && self.0 == *other
+    }
+}
+
+impl TryFrom<Robj> for Rint {
+    type Error = Error;
+
+    fn try_from(robj: Robj) -> Result<Self> {
+        // Check if the value is a scalar
+        match robj.len() {
+            0 => return Err(Error::ExpectedNonZeroLength(robj)),
+            1 => {}
+            _ => return Err(Error::ExpectedScalar(robj)),
+        };
+
+        // Check if the value is not a missing value
+        if robj.is_na() {
+            return Ok(Rint::na());
+        }
+
+        // If the conversion is int-to-int, check the limits. This
+        // needs to be done by `TryFrom` because the conversion by `as`
+        // is problematic when converting a negative value to unsigned
+        // integer types (e.g. `-1i32 as u8` becomes 255).
+        if let Some(v) = robj.as_integer() {
+            if let Ok(v) = Self::try_from(v) {
```

Comment:
I think we can skip this `try_from` in this case. `Rint` is 32-bit, so it should never overflow.

Reply:
Yes.